sdk/dotnet/cmd/pulumi-language-dotnet/main.go

```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
+	ctx context.Context, packageDir, packageName, packageVersion string) (*pulumirpc.PluginDependency, error) {
+
+	logging.V(5).Infof("GetRequiredPlugins: Determining plugin dependency: %v, %v, %v",
+		packageDir, packageName, packageVersion)
+
+	// Check for a `~/.nuget/packages/package_name/package_version/content/version.txt` file.
+
+	versionFilePath := path.Join(packageDir, strings.ToLower(packageName), packageVersion, "content", "version.txt")
+	logging.V(5).Infof("GetRequiredPlugins: version file path: %v", versionFilePath)
+
+	if _, err := os.Stat(versionFilePath); err != nil {
+		if os.IsNotExist(err) {
+			// Pulumi package doesn't contain a version.txt file.  This is not a resource-plugin.
+			// just ignore it.
+			logging.V(5).Infof("GetRequiredPlugins: No such file")
+			return nil, nil
+		}
+
+		// some other error.  report it as it means we can't read this important file.
+		logging.V(5).Infof("GetRequiredPlugins: err: %v", err)
+		return nil, err
+	}
+
+	b, err := ioutil.ReadFile(versionFilePath)
+	if err != nil {
+		logging.V(5).Infof("GetRequiredPlugins: err: %v", err)
+		return nil, err
 	}
 
-	return host.DotnetRun(ctx, req)
+	// Given a package name like "Pulumi.Azure" lowercase the part after Pulumi. to get the plugin name "azure".
+	name := strings.ToLower(packageName[len("Pulumi."):])
+
+	version := strings.TrimSpace(bytes.NewBuffer(b).String())
+	if !strings.HasPrefix(version, "v") {
+		// Version file has stripped off the "v" that we need. So add it back here.
+		version = fmt.Sprintf("v%v", version)
+	}
+
+	result := pulumirpc.PluginDependency{
+		Name:    name,
+		Version: version,
+		Kind:    "resource",
+	}
+
+	logging.V(5).Infof("GetRequiredPlugins: Determining plugin dependency: %#v", result)
+	return &result, nil
 }
 
-func (host *dotnetLanguageHost) DotnetBuild(ctx context.Context, req *pulumirpc.RunRequest) error {
+func (host *dotnetLanguageHost) DotnetBuild(ctx context.Context, req *pulumirpc.GetRequiredPluginsRequest) error {
 	args := []string{"build"}
 
 	if req.GetProgram() != "" {
 		args = append(args, req.GetProgram())
 	}
 
-	if glog.V(5) {
-		commandStr := strings.Join(args, " ")
-		glog.V(5).Infoln("Language host launching process: ", host.exec, commandStr)
+	// Run the `dotnet build` command.  Importantly, report the output of this to the user
+	// (ephemerally) as it is happening so they're aware of what's going on and can see the progress
+	// of things.
+	_, err := host.RunDotnetCommand(ctx, args, true /*logToUser*/)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (host *dotnetLanguageHost) RunDotnetCommand(ctx context.Context, args []string, logToUser bool) (string, error) {
+	commandStr := strings.Join(args, " ")
+	if logging.V(5) {
+		logging.V(5).Infoln("Language host launching process: ", host.exec, commandStr)
 	}
 
 	// Make a connection to the real engine that we will log messages to.
 	conn, err := grpc.Dial(host.engineAddress, grpc.WithInsecure())
```

Comment:
We shouldn't need to do this per-command. We should be able to connect to the engine up-front and keep that connection around.

Reply:
Done.